libfwupdplugin/fu-efivar-freebsd.c

```diff
@@ -0,0 +1,396 @@
+
+/**
+ * fu_efivar_get_path, fu_efivar_get_filename:
+ * Linux-only functions, kept here temporarily for compatibility with
+ * plugins that haven't been adapted yet
+ */
+
+static gchar *
+fu_efivar_get_path (void)
+{
+	return NULL;
+}
+
+static gchar *
+fu_efivar_get_filename (const gchar *guid, const gchar *name)
+{
+	return NULL;
+}
+
+/**
+ * fu_efivar_supported:
+ * @error: #GError
+ *
+ * Determines if the kernel supports EFI variables
+ *
+ * Returns: %TRUE on success
+ *
+ * Since: 1.6.1
+ **/
+gboolean
+fu_efivar_supported (GError **error)
+{
+	if (efi_variables_supported () == 0)
+		return FALSE;
+	return TRUE;
+}
+
+/**
+ * fu_efivar_space_used:
+ * @error: (nullable): optional return location for an error
+ *
+ * Gets the total size used by all EFI variables. This may be less than the size reported by the
+ * kernel as some (hopefully small) variables are hidden from userspace.
+ *
+ * Returns: total allocated size of all visible variables, or %G_MAXUINT64 on error
+ *
+ * Since: 1.6.1
+ **/
+guint64
+fu_efivar_space_used (GError **error)
+{
+	guint64 total = 0;
+	efi_guid_t *guid = NULL;
+	char *name = NULL;
+
+	g_return_val_if_fail (error == NULL || *error == NULL, G_MAXUINT64);
+
+	while (efi_get_next_variable_name (&guid, &name)) {
+		size_t size = 0;
+		if (efi_get_variable_size (*guid, name, &size) < 0)
+			return G_MAXUINT64;
+		total += size;
+	}
+
+	/* success */
+	return total;
+}
+
+/**
+ * fu_efivar_set_data:
+ * @guid: Globally unique identifier
+ * @name: Variable name
+ * @data: Data to set
+ * @sz: size of @data
+ * @attr: Attributes
+ * @error: (nullable): optional return location for an error
+ *
+ * Sets the data to a UEFI variable in NVRAM
+ *
+ * Returns: %TRUE on success
+ *
+ * Since: 1.6.1
+ **/
+gboolean
+fu_efivar_set_data (const gchar *guids, const gchar *name, const guint8 *data,
+		     gsize sz, guint32 attr, GError **error)
+{
+	efi_guid_t guid;
+	efi_str_to_guid (guids, &guid);
+
+	if (efi_set_variable (guid, name, data, sz, attr) != 0) {
+		g_prefix_error (error, "failed to write data to efivar: ");
+		return FALSE;
+	}
+
+	/* success */
+	return TRUE;
+}
+
+/**
+ * fu_efivar_delete:
+ * @guid: Globally unique identifier
+ * @name: Variable name
+ * @error: #GError
+ *
+ * Removes a variable from NVRAM
+ *
+ * Returns: %TRUE on success
+ *
+ * Since: 1.6.1
+ **/
+gboolean
+fu_efivar_delete (const gchar *guids, const gchar *name, GError **error)
+{
+	efi_guid_t guid;
+	efi_str_to_guid (guids, &guid);
+
+	if (efi_del_variable (guid, name) == 0)
+		return TRUE;
+	return FALSE;
+}
+
+/**
+ * fu_efivar_delete_with_glob:
+ * @guid: Globally unique identifier
+ * @name_glob: Variable name
+ * @error: #GError
+ *
+ * Removes a group of variables from NVRAM
+ *
+ * Returns: %TRUE on success
+ *
+ * Since: 1.6.1
+ **/
+gboolean
+fu_efivar_delete_with_glob (const gchar *guids, const gchar *name_glob, GError **error)
+{
+	efi_guid_t guid_to_delete;
+	efi_str_to_guid (guids, &guid_to_delete);
```

Comment:
The call should probably follow all declarations.

Reply:
Fixed.